### PR TITLE
Karma to Jest: graph (refactor)

### PR DESCRIPTION
**Pull request by tskarhed**
Ref #12224 
I did some refactoring to make it more testable, feedback is appreciated! Extracted link function as a class, ran all event stuff in the constructor, made variables into class properties and changed the functions into class methods.

#12809 can probably be merged after this one.

### CODE REVIEW

File: public/app/plugins/panel/graph/graph.ts [hunk 1]
```diff
@@ -21,699 +21,726 @@ import { convertToHistogramData } from './histogram';
...
+  panelWidth: number;
+  eventManager: EventManager;
+  thresholdManager: ThresholdManager;
+
+  constructor(private scope, private elem, timeSrv) {
+    this.ctrl = scope.ctrl;
+    this.dashboard = this.ctrl.dashboard;
+    this.panel = this.ctrl.panel;
+    this.annotations = [];
+    // this.data;
```
Comment (daniellee) [inline public/app/plugins/panel/graph/graph.ts]: Should these commented out lines be removed?

File: public/app/plugins/panel/graph/graph.ts [hunk 2]
```diff
@@ -21,699 +21,726 @@ import { convertToHistogramData } from './histogram';
...
+    // this.sortedSeries;
+    this.panelWidth = 0;
+    this.eventManager = new EventManager(this.ctrl);
+    this.thresholdManager = new ThresholdManager(this.ctrl);
+    this.tooltip = new GraphTooltip(elem, this.ctrl.dashboard, this.scope, () => {
+      return this.sortedSeries;
+    });
+
+    // panel events
+    this.ctrl.events.on('panel-teardown', () => {
```
Comment (daniellee) [inline public/app/plugins/panel/graph/graph.ts]: The event handler functions should be moved out of the constructor and extracted into methods. An example is the heatmap controller: https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/heatmap/heatmap_ctrl.ts#L128-L133

File: public/app/plugins/panel/graph/specs/graph.jest.ts [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+
+describe('grafanaGraph', function() {
+  const setupCtx = (beforeRender?) => {
+    config.bootData = {
+      user: {
+        lightTheme: false,
+      },
+    };
+    GraphCtrl.prototype = <any>{
+      ...MetricsPanelCtrl.prototype,
```
Comment (daniellee) [inline public/app/plugins/panel/graph/specs/graph.jest.ts]: Elegant. I like it!

**Review by daniellee — CHANGES_REQUESTED** [review]
Looks good - just two small improvements that would make it even better.

**Comment by dehrax** [comment]
Changes fixed!

File: public/app/plugins/panel/graph/graph.ts [hunk 4]
```diff
@@ -21,699 +21,730 @@ import { convertToHistogramData } from './histogram';
...
-        updateLegendValues(data, panel, graphHeight);
-
-        ctrl.events.emit('render-legend');
-      });
-
-      ctrl.events.on('legend-rendering-complete', () => {
-        render_panel();
+import { GraphCtrl } from './module';
+
+class Link {
```
Comment (torkelo) [inline public/app/plugins/panel/graph/graph.ts]: Link name feels a bit undescriptive here. Maybe GraphElement?

**Review by torkelo — CHANGES_REQUESTED** [review]
looks good, just the naming of the class Link that feels a bit too generic.

**Pull request merged**